music/src/gansynth/audio_utils.ts

```diff
@@ -0,0 +1,231 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as tf from '@tensorflow/tfjs';
+//@ts-ignore
+import * as FFT from 'fft.js';
+
+import {applyWindow, frame, hannWindow} from './../transcription/audio_utils';
+import {padCenterToLength, SpecParams} from './../transcription/audio_utils';
+//@ts-ignore
+import {MAG_DESCALE_A, MAG_DESCALE_B, N_FFT, N_HOP} from './constants';
+import {PHASE_DESCALE_A, PHASE_DESCALE_B, SAMPLE_RATE} from './constants';
+import {MEL_SPARSE_COEFFS} from './mel_sparse_coeffs';
+
+export function melToLinearMatrix() {
+  const m2l = tf.buffer([1024, 1024]);
+  for (let i = 0; i < MEL_SPARSE_COEFFS.length; i++) {
+    const x = MEL_SPARSE_COEFFS[i] as number[];
+    m2l.set(x[2], x[0], x[1]);
+  }
+  return m2l.toTensor();
+}
+
+function descale(data: tf.Tensor, a: number, b: number) {
+  return tf.div(tf.sub(data, b), a);
+}
+
+export function melToLinear(melLogPower: tf.Tensor3D) {
+  const m2l = melToLinearMatrix().expandDims(0);
+  const melLogPowerDb = descale(melLogPower, MAG_DESCALE_A, MAG_DESCALE_B);
+  // Linear scale the magnitude.
+  const melPower = tf.exp(melLogPowerDb);
+  // Mel to linear frequency scale.
+  const powerLin = tf.matMul(melPower, m2l);
+  // Power to magnitude.
+  const magLin = tf.sqrt(powerLin);
+  return magLin;
+}
+
+export function ifreqToPhase(ifreq: tf.Tensor) {
+  const m2l = melToLinearMatrix().expandDims(0);
+  const ifreqDescale = descale(ifreq, PHASE_DESCALE_A, PHASE_DESCALE_B);
+  // Need to multiply phase by -1.0 to account for conjugacy difference
+  // between tensorflow and librosa/javascript istft.
+  const phase = tf.cumsum(tf.mul(ifreqDescale, Math.PI), 1);
+  const phaseLin = tf.matMul(phase, m2l);
+  return phaseLin;
+}
+
+async function interleaveReIm(real: tf.Tensor, imag: tf.Tensor) {
+  // Combine and add back in the zero DC component
+  let reImBatch = tf.concat([real, imag], 0).expandDims(3);
+  reImBatch = tf.pad(reImBatch, [[0, 0], [0, 0], [1, 0], [0, 0]]);
+
+  // Interleave real and imaginary for javascript ISTFT.
+  // Hack to interleave [re0, im0, re1, im1, ...] with batchToSpace.
+  const crops = [[0, 0], [0, 0]];
+  const reImInterleave =
+      tf.batchToSpaceND(reImBatch, [1, 2], crops).reshape([128, 4096]);
+  // Convert Tensor to a Float32Array[]
+  const reImArray = reImInterleave.dataSync();
+  const reIm = [] as Float32Array[];
+  for (let i = 0; i < 128; i++) {
+    reIm[i] = reImArray.slice(i * 4096, (i + 1) * 4096) as Float32Array;
+  }
+  return reIm;
+}
+
+async function reImToAudio(reIm: Float32Array[]) {
+  const ispecParams = {
+    nFFt: N_FFT,
+    winLength: N_FFT,
+    hopLength: N_HOP,
+    sampleRate: SAMPLE_RATE
+  };
+  return istft(reIm, ispecParams);
+}
+
+export async function specgramsToAudio(specgram: tf.Tensor4D) {
```

Comment:
Memory leak in this function: Needs `tidy` or disposal of tensors.

Reply:
Thanks, yah patching these up helped.